language-plutus-core/docs/Builtins.md

```diff
@@ -0,0 +1,340 @@
+        :: uni `Includes` res
+        => Proxy res -> TypeScheme uni '[] res
+    TypeSchemeArrow
+        :: uni `Includes` arg
+        => Proxy arg -> TypeScheme uni args res -> TypeScheme uni (arg ': args) res
+
+type family FoldArgs args res where
+    FoldArgs '[]           res = res
+    FoldArgs (arg ': args) res = arg -> FoldArgs args res
+
+data BuiltinFunctionMeaning uni =
+    forall args res. BuiltinFunctionMeaning
+        (TypeScheme uni args res)
+        (FoldArgs args res)
+```
+
+`TypeScheme` describes the type of a primitive function: its first argument is the list of types of arguments of that function and its second argument is the type of the result. E.g. `Char -> Bool -> Integer` gets encoded as
+
+```haskell
+charBoolInteger :: (uni `Includes` Char, uni `Includes` Bool) => TypeScheme uni [Char, Bool] Integer`.
+charBoolInteger =
+    Proxy @Char `TypeSchemeArrow`
+    Proxy @Bool `TypeSchemeArrow`
+    TypeSchemeResult (Proxy @Integer)
+```
+
+`FoldArgs` takes the list of types of arguments of a function and the type of the result and folds them back into the type of that function. E.g. `FoldArgs [Char, Bool] Integer` gives `Char -> Bool -> Integer` back.
+
+Finally, `BuiltinFunctionMeaning` is just a `TypeScheme` of a primitive function and the primitive function itself (of type `FoldArgs args res`).
+
+Then when we encounter `AppBuiltinFunction fun args` during evaluation we evaluate `fun` and look up its meaning in the provided environment of `BuiltinFunctionMeaning`s, which allows to retrieve a `TypeScheme`, as well as a Haskell function `f`. Having the `TypeScheme` we know what type each of `args` must have, so after evaluating them to constants we traverse that list of constants and check if the type tag of every constant is the same one as is currently expected from the `TypeScheme` info: if the tags match, then the types they represent must also match and the Haskell type checker sees that ('cause we're using GADTs), so we can type-safely apply `f` to a freshly obtained constant.
+
+The `TypeScheme` of a built-in function also allows to get the Plutus Core type of that function: it's straightforward recursion on `TypeScheme` where each `TypeSchemeArrow` becomes `->` of the target language and each implicit ``uni `Includes` arg`` representing an argument becomes a `BuiltinType (SomeType uni)`, and ``uni `Includes` res`` representing the resulting type also becomes a `BuiltinType (SomeType uni)`.
+
+So having a single `TypeScheme` is enough to adapt the type checker and the evaluator to extensible built-in functions.
+
+## Saturated vs unsaturated built-in functions
+
+There are two ways to have built-in function in the AST of `Term`:
+
+1. either it's saturated application `AppBuiltinFunction BuiltinFunction [Term]` (i.e. the one that we've been using so far) where a built-in function is always applied to the exact number of arguments that it expects, e.g. `AddInteger` has to always be applied to two arguments
```

Comment:
That's not the one we've been using so far, is it?  At the moment we just have `Apply Term Term` and the first `Term` can be a built-in name.